pylint/checkers/unicode.py

```diff
@@ -0,0 +1,701 @@
+            (
+                "Moves the cursor back, so the character after it will overwrite the "
+                "character before"
+            ),
+        ),
+        _BadChar(
+            "carriage-return",
+            "\r",
+            "\\r",
+            "E2411",
+            (
+                "Moves the cursor to the start of line, subsequent characters overwrite "
+                "the start of the line."
+            ),
+        ),
+        _BadChar(
+            "sub",
+            "\x1A",
+            "\\x1A",
+            "E2412",
+            (
+                "Ctrl+Z “End of text” on Windows. Some programs (such as type) ignore "
+                "the rest of the file after it."
+            ),
+        ),
+        _BadChar(
+            "esc",
+            "\x1B",
+            "\\x1B",
+            "E2413",
+            (
+                "Commonly initiates escape codes which allow arbitrary control "
+                "of the terminal."
+            ),
+        ),
+        _BadChar(
+            "nul",
+            "\0",
+            "\\0",
+            "E2414",
+            "Mostly end of input for python.",
+        ),
+        _BadChar(
+            # Zero Width with Space. At the time of writing not accepted by Python.
+            # But used to within Trojan Source Examples, so still included and tested for.
+            "zero-width-space",
+            "\u200B",  # \n{ZERO WIDTH SPACE}
+            "\\u200B",
+            "E2415",
+            "Invisible space character could hide real code execution.",
+        ),
+    ]
+    _BAD_ASCII_SEARCH_DICT = {char.unescaped: char for char in _BAD_CHARS}
+
+    msgs = {
+        "E2401": (
+            # This error will be only displayed to users once Python Supports
+            # UTF-16/UTF-32 (if at all)
+            "UTF-8 is the recommended Unicode codec.",
+            "invalid-unicode-codec",
+            (
+                "For compatibility use UTF-8 instead of UTF-16/UTF-32. "
+                "See also https://bugs.python.org/issue1503789 for a history"
+                "of this issue. And "
+                "https://softwareengineering.stackexchange.com/questions/102205/should-utf-16-be-considered-harmful"
+                "for some possible problems when using UTF-16 for instance."
+            ),
+        ),
+        "E2402": (
+            "bidirectional unicode can be used to hide real code execution",
+            "bidirectional-unicode",
+            "bidirectional unicode can be used to hide real code execution",
+        ),
+        **{
+            bad_char.code: (
+                bad_char.description(),
+                bad_char.human_code(),
+                bad_char.help_text,
+            )
+            for bad_char in _BAD_CHARS
+        },
+    }
+
+    @staticmethod
+    def _is_invalid_codec(codec: Optional[str]) -> bool:
+        if codec is None:
+            return False
+        codec = codec.lower().replace("-", "").replace("_", "")
```

Comment:
I think we don't need to replace here. Again, I wonder if we can also remove `lower` as well.

Reply:
again replaced with `_normalize_codec_name`

Reply:
Centralized _normalize_codec_name so this is not up to date anymore